dandi/dandiapi.py

```diff
@@ -301,25 +310,87 @@ def paginate(
+        (default 5) at a time.  This behavior requires the initial response to
+        contain a ``"count"`` key giving the number of items across all pages.
+
+        :param Optional[int] page_size:
+            If non-`None`, overrides the client's `page_size` attribute for
+            this sequence of pages
         """
+        if page_size is None:
+            page_size = self.page_size
         if page_size is not None:
             if params is None:
                 params = {}
             params["page_size"] = page_size
-        r = self.get(path, params=params, **kwargs)
-        while True:
-            for item in r["results"]:
-                yield item
-            if r.get("next"):
-                r = self.get(r["next"])
-            else:
-                break
+
+        resp = self.get(path, params=params, json_resp=False)
+        r = resp.json()
+        if r["next"] is not None:
+            page1 = resp.history[0].url if resp.history else resp.url
+            if not is_page2_url(page1, r["next"]):
+                lgr.warning(
+                    "Pagination request to %s returned unexpected 'next' URL: %s",
+                    page1,
+                    r["next"],
+                )
+                if os.environ.get("DANDI_PAGINATION_DISABLE_FALLBACK"):
+                    raise RuntimeError(
+                        f"API server changed pagination strategy: {page1} URL"
+                        f" is now followed by {r['next']}"
+                    )
+                else:
+                    while True:
+                        yield from r["results"]
+                        if r.get("next"):
+                            r = self.get(r["next"])
+                        else:
+                            return
+        yield from r["results"]
+        if r["next"] is None:
+            return
+
+        if page_size is None:
+            page_size = len(r["results"])
+        pages = (r["count"] + page_size - 1) // page_size
+        pageno_iter = iter(range(2, pages + 1))
+
+        with ThreadPoolExecutor(max_workers=self.page_lookahead) as pool:
+            futures: deque[Future[list]] = deque()
+
+            def get_page(pageno: int) -> list:
+                params2 = params.copy() if params is not None else {}
+                params2["page"] = pageno
+                results = self.get(path, params=params2)["results"]
+                assert isinstance(results, list)
+                return results
+
+            def submit_job() -> None:
+                try:
+                    i = next(pageno_iter)
+                except StopIteration:
+                    return
+                futures.append(pool.submit(get_page, i))
+
+            try:
+                for _ in range(self.page_lookahead):
+                    submit_job()
+                while futures:
+                    res = futures.popleft().result()
+                    submit_job()
+                    yield from res
```

Comment:
I guess I was not specific enough in https://github.com/dandi/dandi-cli/pull/1021#discussion_r883959010 . What I meant: If I got your investigation right in https://github.com/dandi/dandi-cli/pull/1021#discussion_r883936703 - we will reach `finally` in any case, even though may be slightly delayed waiting for `gc`  to kick in, so I think that simpler/analogous to other places 

```suggestion
            try:
                for i in range(2, pages + 1):        
                    futures.append(pool.submit(get_page, i))
                while futures:
                    yield futures.popleft().result()
```
would be just fine (would also eliminate that definition of `pageno_iter` above) or what am I still missing?

Reply:
Submitting everything to the pool at once could still result in far more pages being fetched than are needed, e.g., if the user iterates through the results slowly (so that there's enough time for dozens or hundreds of pages to be fetched in the background) and still ends up breaking out of the loop early on.

Reply:
Well, I would then use your argument of "slow iteration" that what if user iterates slowly for awhile (laptop/bandwidth/server is busy with something else) and then really fast: Your approach would cause the the slow down of fetching during "slow" phase, instead of fetching as fast as it can to provide results during "fast" phase.

Overall, if we had to deal with such effects, other code paths of similar kind should have got similar code refactoring but I truly do not think it is necessary.  The goal for this function is to fetch most efficiently, so let's stick to that and a uniform code pattern.

Reply:
Thank you @jwodder !